File: src/main/java/com/ljc/workprogress/ui/editor/preview/WpsStaticServer.java
package com.ljc.workprogress.ui.editor.preview;

import com.google.common.base.Joiner;
import com.google.common.collect.Maps;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.application.ApplicationInfo;
import com.intellij.util.Url;
import com.intellij.util.Urls;
import io.netty.buffer.Unpooled;
import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.http.*;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.math.NumberUtils;
import org.jetbrains.ide.BuiltInServerManager;
import org.jetbrains.ide.HttpRequestHandler;
import org.jetbrains.io.FileResponses;
import org.jetbrains.io.Responses;

import java.util.Date;
import java.util.HashMap;

public class WpsStaticServer extends HttpRequestHandler {

    private static final String prefixUuid = "da2f7518-aefb-4187-9792-574c65473521";
    private static final String prefixPath = "/" + prefixUuid;

    private static ResourceProvider defaultResourceProvider = new DefaultResourceProvider();
    private static HashMap<Integer, ResourceProvider> resourceProviders = Maps.newHashMap();

    static {
        resourceProviders.put(defaultResourceProvider.hashCode(), defaultResourceProvider);
    }

    public static WpsStaticServer getInstance() {
        return WpsStaticServer.Companion.getEP_NAME().findExtension(WpsStaticServer.class);
    }

    public static String getStaticUrl(ResourceProvider resourceProvider, String staticPath) {
        int providerHash = resourceProvider.hashCode();
        int port = BuiltInServerManager.getInstance().getPort();
        String raw = String.format("http://localhost:%s/%s/%s/%s", port, prefixUuid, providerHash, staticPath);
        Url url = Urls.parseEncoded(raw);
        if (url == null) {
            throw new RuntimeException("Could not parse url!");
        }
        return BuiltInServerManager.getInstance().addAuthToken(url).toExternalForm();
    }

    public synchronized Disposable registerResourceProvider(ResourceProvider resourceProvider) {
        resourceProviders.put(resourceProvider.hashCode(), resourceProvider);
        return new Disposable() {
            @Override
            public void dispose() {
                unregisterResourceProvider(resourceProvider);
            }
        };
    }

    public synchronized void unregisterResourceProvider(ResourceProvider resourceProvider) {
        resourceProviders.remove(resourceProvider.hashCode());
    }

    private Integer getProviderHash(String path) {
        String[] split = path.split("/");
        if (split.length >= 2) {
            return NumberUtils.toInt(split[2]);
        }
        return null;
    }

    private String getStaticPath(String path) {

        String[] split = path.split("/");
        if (split.length >= 3) {
            String[] arr = ArrayUtils.subarray(split, 3, split.length);
            return Joiner.on("/").join(arr);

        }
        return "";
    }

    private ResourceProvider obtainResourceProvider(String path) {
        Integer providerHash = getProviderHash(path);
        synchronized (resourceProviders) {
            return resourceProviders.getOrDefault(providerHash, null);
        }

    }

    @Override
    public boolean isSupported(FullHttpRequest request) {
        if (!super.isSupported(request)) {
            return false;
        }
        String path = request.uri();
        return path.startsWith(prefixPath);
    }

    @Override
    public boolean process(QueryStringDecoder urlDecoder, FullHttpRequest request, ChannelHandlerContext context) {
        String path = urlDecoder.path();

        ResourceProvider resourceProvider = obtainResourceProvider(path);
        if (resourceProvider == null) {
            return false;
        }
        String resourceName = getStaticPath(path);
        if (resourceProvider.canProvide(resourceName)) {
            sendResource(request, context.channel(), resourceProvider.loadResource(resourceName), resourceName);
            return true;
        }
        return false;
    }

    private String[] typesForExplicitUtfCharset = new String[]{"application/javascript", "text/html", "text/css", "image/svg+xml"};

    private String guessContentType(String resourceName) {
        String type = FileResponses.INSTANCE.getContentType(resourceName);
        if (ArrayUtils.contains(typesForExplicitUtfCharset, type)) {
            return String.format("%s; charset=utf-8", type);
        }
        return type;
    }

    private void sendResource(HttpRequest request, Channel channel, Resource resource, String resourceName) {
        long lastModified = ApplicationInfo.getInstance().getBuildDate().getTimeInMillis();

        if (FileResponses.INSTANCE.checkCache(request, channel, lastModified)) {
            return;
        }
        if (resource == null) {
            Responses.send(HttpResponseStatus.NOT_FOUND, channel, request);
            return;
        }
        FullHttpResponse response = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK, Unpooled.wrappedBuffer(resource.getContent()));
        if (resource.getType() != null) {
            response.headers().set(HttpHeaderNames.CONTENT_TYPE, resource.getType());
        } else {
            response.headers().set(HttpHeaderNames.CONTENT_TYPE, guessContentType(resourceName));
        }
        response.headers().set(HttpHeaderNames.CACHE_CONTROL, "private, must-revalidate");
        response.headers().set(HttpHeaderNames.LAST_MODIFIED, new Date(lastModified));
        Responses.send(response, channel, request);

    }
}

File: src/main/kotlin/com/jdl/ljc/joyworkprogress/WpsStaticServer.kt
package com.jdl.ljc.joyworkprogress

import com.intellij.openapi.Disposable
import com.intellij.openapi.application.ApplicationInfo
import com.intellij.openapi.util.text.StringUtil
import com.intellij.util.Urls
import io.netty.buffer.Unpooled
import io.netty.channel.Channel
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.http.*
import org.jetbrains.ide.BuiltInServerManager
import org.jetbrains.ide.HttpRequestHandler
import org.jetbrains.io.FileResponses
import org.jetbrains.io.send
import java.net.URL
import java.util.*

/**
 *
 * @author liangjichao
 * @date 2023/10/27 3:42 PM
 */
class WpsStaticServer : HttpRequestHandler() {
    private val defaultResourceProvider = ResourceProvider.DefaultResourceProvider()
    private val resourceProviders = hashMapOf<Int, ResourceProvider>(defaultResourceProvider.hashCode() to defaultResourceProvider)

    /**
     * @return [Disposable] which will unregister [resourceProvider].
     */
    @Synchronized
    fun registerResourceProvider(resourceProvider: ResourceProvider): Disposable {
        resourceProviders[resourceProvider.hashCode()] = resourceProvider
        return Disposable { unregisterResourceProvider(resourceProvider) }
    }

    /**
     * Prefer unregistering providers by disposing [Disposable] returned by [registerResourceProvider].
     */
    @Synchronized
    fun unregisterResourceProvider(resourceProvider: ResourceProvider) {
        resourceProviders.remove(resourceProvider.hashCode())
    }

    private fun getProviderHash(path: String): Int? {
        return path.split('/').getOrNull(2)?.toIntOrNull()
    }

    private fun getStaticPath(path: String): String {
        return path.split('/').drop(3).joinToString(separator = "/")
    }

    private fun obtainResourceProvider(path: String): ResourceProvider? {
        val providerHash = getProviderHash(path) ?: return null
        return synchronized(resourceProviders) { resourceProviders.getOrDefault(providerHash, null) }
    }

    override fun isSupported(request: FullHttpRequest): Boolean {
        if (!super.isSupported(request)) {
            return false
        }
        val path = request.uri()
        return path.startsWith(prefixPath)
    }

    override fun process(urlDecoder: QueryStringDecoder, request: FullHttpRequest, context: ChannelHandlerContext): Boolean {
        val path = urlDecoder.path()
        check(path.startsWith(prefixPath)) { "prefix should have been checked by #isSupported" }
        val resourceProvider = obtainResourceProvider(path) ?: return false
        val resourceName = getStaticPath(path)
        if (resourceProvider.canProvide(resourceName)) {
            sendResource(request, context.channel(), resourceProvider.loadResource(resourceName), resourceName)
            return true
        }
        return false
    }

    companion object {
        private const val prefixUuid = "da2f7518-aefb-4187-9792-574c65473521"
        private const val prefixPath = "/$prefixUuid"

        @JvmStatic
        val instance: WpsStaticServer
            get() = EP_NAME.findExtension(WpsStaticServer::class.java) ?: error("Could not get server instance!")

        @JvmStatic
        fun createCSP(scripts: List<String>, styles: List<String>): String {
            // We need to remove any query parameters to stop annoying errors in the browser console
            fun stripQueryParameters(url: String) = url.replace("?${URL(url).query}", "")
            return """
        default-src 'none';
        script-src ${StringUtil.join(scripts.map(::stripQueryParameters), " ")};
        style-src https: ${StringUtil.join(styles.map(::stripQueryParameters), " ")} 'unsafe-inline';
        img-src file: *; connect-src 'none'; font-src * data: *;
        object-src 'none'; media-src 'none'; child-src 'none';
      """
        }

        /**
         * Expected to return same URL on each call for same [resourceProvider] and [staticPath],
         * if [resourceProvider] was not unregistered between those calls.
         */
        @JvmStatic
        fun getStaticUrl(resourceProvider: ResourceProvider, staticPath: String): String {
            val providerHash = resourceProvider.hashCode()
            val port = BuiltInServerManager.getInstance().port
            val raw = "http://localhost:$port/$prefixUuid/$providerHash/$staticPath"
            val url = Urls.parseEncoded(raw)
            requireNotNull(url) { "Could not parse url!" }
            return BuiltInServerManager.getInstance().addAuthToken(url).toExternalForm()
        }

        @Deprecated("Use PreviewStaticServer.getStaticUrl(ResourceProvider, String) instead")
        @JvmStatic
        fun getStaticUrl(staticPath: String): String {
            return getStaticUrl(instance.defaultResourceProvider, staticPath)
        }

        /**
         * The types for which `";charset=utf-8"` will be appended (only if guessed by [guessContentType]).
         */
        private val typesForExplicitUtfCharset = arrayOf("application/javascript", "text/html", "text/css", "image/svg+xml")

        private fun guessContentType(resourceName: String): String {
            val type = FileResponses.getContentType(resourceName)
            return if (type in typesForExplicitUtfCharset) {
                "$type; charset=utf-8"
            } else type
        }

        private fun sendResource(request: HttpRequest, channel: Channel, resource: ResourceProvider.Resource?, resourceName: String) {
            val lastModified = ApplicationInfo.getInstance().buildDate.timeInMillis
            if (FileResponses.checkCache(request, channel, lastModified)) {
                return
            }
            if (resource == null) {
                HttpResponseStatus.NOT_FOUND.send(channel, request)
                return
            }
            val response: FullHttpResponse = DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK, Unpooled.wrappedBuffer(resource.content))
            with(response) {
                if (resource.type != null) {
                    headers()[HttpHeaderNames.CONTENT_TYPE] = resource.type
                } else {
                    headers()[HttpHeaderNames.CONTENT_TYPE] = guessContentType(resourceName)
                }
                headers()[HttpHeaderNames.CACHE_CONTROL] = "private, must-revalidate"
                headers()[HttpHeaderNames.LAST_MODIFIED] = Date(lastModified)
                send(channel, request)
            }
        }
    }
}